### PR TITLE
AVRO-4182: Avoid ANTLR parsing edge cases

### PR DESCRIPTION
**Pull request by opwvhk**
## What is the purpose of the change

In some cases, enter and exit rules are not called in pairs. Apparently, this is by design. This fix adds a test that the fix does indeed work (it tests an exception, as the error does not show up with a duplicated definition).

Note: this change keeps the entire parse tree in memory during parsing, and allows it to be collected directly afterwards.

This PR fixes the bug reported as AVRO-4182.

## Verifying this change

This change adds a test: `org.apache.avro.idl.TestIdlReader#testErrorDuplicateNamedType()`

## Documentation

- Does this pull request introduce a new feature? (~yes~ / **no**)
- If yes, how is the feature documented? (**not applicable** / ~docs / JavaDocs / not documented~)

### CODE REVIEW

File: lang/java/idl/src/main/java/org/apache/avro/idl/IdlReader.java [hunk 1]
```diff
@@ -207,20 +208,22 @@ private IdlFile parse(URI inputDir, CharStream charStream) {
     IdlParser parser = new IdlParser(tokenStream);
     parser.removeErrorListeners();
     parser.addErrorListener(SIMPLE_AVRO_ERROR_LISTENER);
-    parser.addParseListener(parseListener);
+    // parser.addParseListener(parseListener);
```
Comment (nandorKollar) [inline lang/java/idl/src/main/java/org/apache/avro/idl/IdlReader.java]: I think you can remove this line.

**Comment by opwvhk** [comment]
Cherry-picked to 1.12
